OkHttpClient connection leak in OperationSupport / HttpResponse

Hi,

while checking our logs we observed a warning from the OkHttpClient that a connection was leaked. As you can see in the stack trace at the very bottom of this ticket our application tried to call BaseOperation.list(), but ran into an unauthorized error. Shortly after that the OkHttpClient reports a connection leak.

Looking at your code we can see that in case of an unsuccessful response [OperationSupport.assertResponseCode()](https://github.com/fabric8io/kubernetes-client/blob/b5231ae8407493cd7cd45499c29bdbb2d8a5057a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/OperationSupport.java#L640) calls [OperationSupport.createStatus()](https://github.com/fabric8io/kubernetes-client/blob/b5231ae8407493cd7cd45499c29bdbb2d8a5057a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/OperationSupport.java#L673) which in turn gets the HTTP body string by calling [HttpResponse.bodyString()](https://github.com/fabric8io/kubernetes-client/blob/b5231ae8407493cd7cd45499c29bdbb2d8a5057a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/OperationSupport.java#L680). The default implementation of bodyString() of the HttpResponse interface then gets one of the [SupportedResponses](https://github.com/fabric8io/kubernetes-client/blob/b5231ae8407493cd7cd45499c29bdbb2d8a5057a/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/http/HttpResponse.java#L68).  In case of an input stream or reader this then calls IOHelpers.readFully. However the readFully does not close the stream that it is reading from. Our assumption is that in case of an error the SupportedResponses.INPUT_STREAM or SupportedResponses.READER is used. When looking at the [OkHttpResponseImpl](https://github.com/fabric8io/kubernetes-client/blob/b5231ae8407493cd7cd45499c29bdbb2d8a5057a/httpclient-okhttp/src/main/java/io/fabric8/kubernetes/client/okhttp/OkHttpClientImpl.java#L167) we see that the response body is closed in one case explicitly and for the string() and bytes() calls it is closed implicitly (according to the Kotlin sources of OkHttp). However, for the charStream() and byteStream() the response body should probably be closed be the consumer of the stream which is not happening in the SupportedResponses. And therefore the connection is leaked.

```
2022-12-31 22:08:19.788 WARN  [okhttp3.OkHttpClient] A connection to https://172.20.0.1/ was leaked. Did you forget to close a response body? To see where this was allocated, set the OkHttpClient logger level to FINE: Logger.getLogger(OkHttpClient.class.getName()).setLevel(Level.FINE);

2022-12-31 22:03:18.952 [...] WARN io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.20.0.1:443/api/v1/nodes?labelSelector=test%2Ftier%3Dapp. Message: Unauthorized! Configured service account doesn't have access. Service account may have been revoked. Unauthorized.
	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:238) ~[kubernetes-client-api-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:536) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.list(BaseOperation.java:418) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.list(BaseOperation.java:383) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.list(BaseOperation.java:93) ~[kubernetes-client-6.3.1.jar:?]
[...]
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.20.0.1:443/api/v1/nodes?labelSelector=test%2Ftier%3Dapp. Message: Unauthorized! Configured service account doesn't have access. Service account may have been revoked. Unauthorized.
at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:742) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:722) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:671) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:601) ~[kubernetes-client-6.3.1.jar:?]
[...]
```



### Fabric8 Kubernetes Client version

6.3.1

### Steps to reproduce

1. Call BaseOperation.list()
2. Let the server return an unauthorized response for your request with the corresponding status code triggering the client to call OperationSupport.createStatus()
3. Logs show a leaked OkHttp connection shortly afterwards

### Expected behavior

No connection leak warning from OkHttp

### Runtime

other (please specify in additional context)

### Kubernetes API Server version

other (please specify in additional context)

### Environment

Amazon

### Fabric8 Kubernetes Client Logs

```shell
2022-12-31 22:08:19.788 WARN  [okhttp3.OkHttpClient] A connection to https://172.20.0.1/ was leaked. Did you forget to close a response body? To see where this was allocated, set the OkHttpClient logger level to FINE: Logger.getLogger(OkHttpClient.class.getName()).setLevel(Level.FINE);

2022-12-31 22:03:18.952 [...] WARN io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.20.0.1:443/api/v1/nodes?labelSelector=test%2Ftier%3Dapp. Message: Unauthorized! Configured service account doesn't have access. Service account may have been revoked. Unauthorized.
	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:238) ~[kubernetes-client-api-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:536) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.list(BaseOperation.java:418) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.list(BaseOperation.java:383) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.list(BaseOperation.java:93) ~[kubernetes-client-6.3.1.jar:?]
[...]
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.20.0.1:443/api/v1/nodes?labelSelector=test%2Ftier%3Dapp. Message: Unauthorized! Configured service account doesn't have access. Service account may have been revoked. Unauthorized.
at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:742) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:722) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:671) ~[kubernetes-client-6.3.1.jar:?]
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:601) ~[kubernetes-client-6.3.1.jar:?]
[...]
```


### Additional context

Unfortunately I don't know the kubernetes flavor or API server version, but it is anyways not relevant for this issue.

--- a/httpclient-okhttp/pom.xml
+++ b/httpclient-okhttp/pom.xml
@@ -81,7 +81,7 @@
     </dependency>
     <dependency>
       <groupId>org.mockito</groupId>
-      <artifactId>mockito-core</artifactId>
+      <artifactId>mockito-inline</artifactId>
     </dependency>
     <dependency>
       <groupId>io.fabric8</groupId>

--- a/httpclient-okhttp/src/main/java/io/fabric8/kubernetes/client/okhttp/OkHttpClientImpl.java
+++ b/httpclient-okhttp/src/main/java/io/fabric8/kubernetes/client/okhttp/OkHttpClientImpl.java
@@ -159,7 +159,7 @@ public class OkHttpClientImpl extends StandardHttpClient<OkHttpClientImpl, OkHtt
     private T body;
     private Class<T> type;
 
-    public OkHttpResponseImpl(Response response, T body) throws IOException {
+    public OkHttpResponseImpl(Response response, T body) {
       this.response = response;
       this.body = body;
     }

--- a/httpclient-okhttp/src/main/java/io/fabric8/kubernetes/client/okhttp/OkHttpWebSocketImpl.java
+++ b/httpclient-okhttp/src/main/java/io/fabric8/kubernetes/client/okhttp/OkHttpWebSocketImpl.java
@@ -81,6 +81,7 @@ class OkHttpWebSocketImpl implements WebSocket {
           if (response != null) {
             try {
               future.complete(new WebSocketResponse(null,
+                  // passing null as the type ensures that the response body is closed
                   new WebSocketHandshakeException(new OkHttpResponseImpl<>(response, null)).initCause(t)));
             } catch (IOException e) {
               // can't happen

--- a/httpclient-okhttp/src/test/java/io/fabric8/kubernetes/client/okhttp/OkHttpResponseImplTest.java
+++ b/httpclient-okhttp/src/test/java/io/fabric8/kubernetes/client/okhttp/OkHttpResponseImplTest.java
@@ -1,0 +1,41 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.kubernetes.client.okhttp;
+
+import okhttp3.Response;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+
+import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+class OkHttpResponseImplTest {
+
+  @Test
+  @DisplayName("constructor, with type=null, should close the response body")
+  void nullTypeClosesResponseBody() throws IOException {
+    // Given
+    final Response response = mock(Response.class, RETURNS_DEEP_STUBS);
+    // When
+    new OkHttpClientImpl.OkHttpResponseImpl<>(response, null);
+    // Then
+    verify(response.body(), times(1)).close();
+  }
+}

--- a/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/http/StandardHttpClient.java
+++ b/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/http/StandardHttpClient.java
@@ -63,6 +63,8 @@ public abstract class StandardHttpClient<C extends HttpClient, F extends HttpCli
               .afterFailure(copy, response)
               .thenCompose(b -> {
                 if (Boolean.TRUE.equals(b)) {
+                  // before starting another request, make sure the old one is cancelled / closed
+                  response.body().cancel();
                   return consumeBytesDirect(copy.build(), consumer);
                 }
                 return CompletableFuture.completedFuture(response);

--- a/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/http/AbstractInterceptorTest.java
+++ b/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/http/AbstractInterceptorTest.java
@@ -27,8 +27,10 @@ import java.nio.charset.StandardCharsets;
 import java.util.Collections;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public abstract class AbstractInterceptorTest {
 
@@ -146,11 +148,13 @@ public abstract class AbstractInterceptorTest {
   public void afterHttpFailureReplacesResponseInSendAsync() throws Exception {
     // Given
     server.expect().withPath("/intercepted-url").andReturn(200, "This works").once();
+    AtomicReference<HttpResponse<?>> responseRef = new AtomicReference<>();
     final HttpClient.Builder builder = getHttpClientFactory().newBuilder()
         .addOrReplaceInterceptor("test", new Interceptor() {
           @Override
           public CompletableFuture<Boolean> afterFailure(BasicBuilder builder, HttpResponse<?> response) {
             builder.uri(URI.create(server.url("/intercepted-url")));
+            responseRef.set(response);
             return CompletableFuture.completedFuture(true);
           }
         });
@@ -163,6 +167,8 @@ public abstract class AbstractInterceptorTest {
       assertThat(result)
           .returns("This works", HttpResponse::body)
           .returns(200, HttpResponse::code);
+
+      assertTrue(((AsyncBody) responseRef.get().body()).done().isDone());
     }
   }
 
@@ -171,11 +177,13 @@ public abstract class AbstractInterceptorTest {
   public void afterHttpFailureReplacesResponseInConsumeBytes() throws Exception {
     // Given
     server.expect().withPath("/intercepted-url").andReturn(200, "This works").once();
+    AtomicReference<HttpResponse<?>> responseRef = new AtomicReference<>();
     final HttpClient.Builder builder = getHttpClientFactory().newBuilder()
         .addOrReplaceInterceptor("test", new Interceptor() {
           @Override
           public CompletableFuture<Boolean> afterFailure(BasicBuilder builder, HttpResponse<?> response) {
             builder.uri(URI.create(server.url("/intercepted-url")));
+            responseRef.set(response);
             return CompletableFuture.completedFuture(true);
           }
         });
@@ -193,6 +201,8 @@ public abstract class AbstractInterceptorTest {
       asyncR.body().done().get(10L, TimeUnit.SECONDS);
       // Then
       assertThat(result.get()).isEqualTo("This works");
+
+      assertTrue(((AsyncBody) responseRef.get().body()).done().isDone());
     }
   }
 

--- a/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/http/AbstractInterceptorTest.java
+++ b/kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/http/AbstractInterceptorTest.java
@@ -25,12 +25,12 @@ import org.junit.jupiter.api.Test;
 import java.net.URI;
 import java.nio.charset.StandardCharsets;
 import java.util.Collections;
+import java.util.Set;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicReference;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public abstract class AbstractInterceptorTest {
 
@@ -148,13 +148,11 @@ public abstract class AbstractInterceptorTest {
   public void afterHttpFailureReplacesResponseInSendAsync() throws Exception {
     // Given
     server.expect().withPath("/intercepted-url").andReturn(200, "This works").once();
-    AtomicReference<HttpResponse<?>> responseRef = new AtomicReference<>();
     final HttpClient.Builder builder = getHttpClientFactory().newBuilder()
         .addOrReplaceInterceptor("test", new Interceptor() {
           @Override
           public CompletableFuture<Boolean> afterFailure(BasicBuilder builder, HttpResponse<?> response) {
             builder.uri(URI.create(server.url("/intercepted-url")));
-            responseRef.set(response);
             return CompletableFuture.completedFuture(true);
           }
         });
@@ -167,8 +165,6 @@ public abstract class AbstractInterceptorTest {
       assertThat(result)
           .returns("This works", HttpResponse::body)
           .returns(200, HttpResponse::code);
-
-      assertTrue(((AsyncBody) responseRef.get().body()).done().isDone());
     }
   }
 
@@ -177,13 +173,11 @@ public abstract class AbstractInterceptorTest {
   public void afterHttpFailureReplacesResponseInConsumeBytes() throws Exception {
     // Given
     server.expect().withPath("/intercepted-url").andReturn(200, "This works").once();
-    AtomicReference<HttpResponse<?>> responseRef = new AtomicReference<>();
     final HttpClient.Builder builder = getHttpClientFactory().newBuilder()
         .addOrReplaceInterceptor("test", new Interceptor() {
           @Override
           public CompletableFuture<Boolean> afterFailure(BasicBuilder builder, HttpResponse<?> response) {
             builder.uri(URI.create(server.url("/intercepted-url")));
-            responseRef.set(response);
             return CompletableFuture.completedFuture(true);
           }
         });
@@ -201,9 +195,44 @@ public abstract class AbstractInterceptorTest {
       asyncR.body().done().get(10L, TimeUnit.SECONDS);
       // Then
       assertThat(result.get()).isEqualTo("This works");
-
-      assertTrue(((AsyncBody) responseRef.get().body()).done().isDone());
     }
+  }
+
+  @Test
+  @DisplayName("afterFailure (HTTP), previous consumed response bodies are closed")
+  public void afterHttpFailurePreviousResponsesAreClosed() throws Exception {
+    // Given
+    server.expect().withPath("/intercepted-url").andReturn(200, "This works").once();
+    final Set<HttpResponse<?>> interceptedResponses = ConcurrentHashMap.newKeySet();
+    final HttpClient.Builder builder = getHttpClientFactory().newBuilder()
+        .addOrReplaceInterceptor("invalid-url", new Interceptor() {
+          @Override
+          public CompletableFuture<Boolean> afterFailure(BasicBuilder builder, HttpResponse<?> response) {
+            builder.uri(URI.create(server.url("/still-not-found")));
+            interceptedResponses.add(response);
+            return CompletableFuture.completedFuture(true);
+          }
+        })
+        .addOrReplaceInterceptor("valid-url", new Interceptor() {
+          @Override
+          public CompletableFuture<Boolean> afterFailure(BasicBuilder builder, HttpResponse<?> response) {
+            builder.uri(URI.create(server.url("/intercepted-url")));
+            interceptedResponses.add(response);
+            return CompletableFuture.completedFuture(true);
+          }
+        });
+    // When
+    try (HttpClient client = builder.build()) {
+      client.consumeBytes(
+          client.newHttpRequestBuilder().uri(server.url("/not-found")).build(),
+          (s, ab) -> ab.consume())
+          .get(10, TimeUnit.SECONDS);
+    }
+    // Then
+    assertThat(interceptedResponses)
+        .hasSize(2)
+        .extracting(r -> ((AsyncBody) r.body()).done().isDone())
+        .containsOnly(true, true);
   }
 
   @Test

